File: DSA-C++/6-array/10-find-missing-number.cpp
#include <bits/stdc++.h>

using namespace std;
// Given an array of size n-1 with distinct integers in the range of [1, n]. This array represents a permutation of the integers from 1 to n with one element missing. Find the missing element in the array.


// brute force - O(n²), O(1)
int missing_number_1(vector <int> &nums) {
    int n = nums.size();

    for (int i=1; i<=n; ++i) {
        bool present = false;

        for (int j=0; j<n; ++j) {
            if (nums[j] == i) {
                present = true;
                break;
            }
        }

        if (!present) {
            return i;
        }
    }

    return n+1;
}


// hashing - O(n), O(n)
int missing_number_2(vector <int> &nums) {
    int n = nums.size();
    unordered_set <int> uset;

    for (int &x:nums) {
        uset.insert(x);
    }

    for (int i=1; i<=n;++i) {
        if (!uset.count(i)) {
            return i;
        }
    }

    return n+1;
}


// sum difference - O(n), O(1)
int missing_number_3(vector <int> &nums) {
    int n = nums.size();
    long long sum2 = (1LL * (n+1) * (n+2)) / 2, sum1=0;

    for (int &x:nums) {
        sum1 += 1LL * x;
    }

    return sum2 - sum1;
}


// xor - O(n), O(1)
int missing_number_4(vector <int> &nums) {
    int n = nums.size() + 1;
    int xor_n = 0, xor_arr = 0;

    for (int &x:nums) {
        xor_arr ^= x;
    }

    for (int i=0; i<=n; ++i) {
        xor_n ^= i;
    }

    return xor_n ^ xor_arr;
}


int main() {
    vector <int> nums = {8, 2, 4, 5, 3, 7, 1};
    nums = {5,1,4,2};
    // nums = {2};

    cout << missing_number_1(nums) << endl;

    cout << missing_number_2(nums) << endl;

    cout << missing_number_3(nums) << endl;

    cout << missing_number_4(nums) << endl;
    
    return 0;
}

File: DSA-Python/5-array/10-find-missing-number.py
# Given an array nums containing n distinct numbers in the range [0, n], return the only number in the range that is missing from the array.



# brute force - O(n²), O(1)
def find_missing_number_1(nums:list) -> int:
    n = len(nums)
    j = 1
    
    for cnt in range(n+1):
        present = False
        
        for num in nums:
            if num == cnt:
                present = True
                break
            
        if not present:
            return cnt
        
    return -1



# hashing - O(n), O(n)
def find_missing_number_2(nums:list) -> int:
    n = len(nums)
    uset = set()
    
    for x in nums:
        uset.add(x)
        
    for i in range(n+1):
        if i not in uset:
            return i
        
    return -1



# sum difference - O(n), O(1)
def find_missing_number_3(nums:list) -> int:
    n = len(nums)
    
    return ((n * (n+1)) // 2) - sum(nums)

    

if __name__=="__main__":
    loop = int(input())
    
    for _ in range(loop):
        nums = list(map(int, input().split()))
        
        res = find_missing_number_1(nums)
        
        res = find_missing_number_2(nums)
        
        res = find_missing_number_3(nums)
        
        print(res)
        
        
'''
5
0 1
9 6 4 2 3 5 7 0 1
1

3 0 1
'''